### ISSUE TITLE
RNN-Direct / LSTM-Direct produce incorrect multi-step forecasts

### ISSUE THREAD

**Issue by alireza-jafari** [Nixtla/neuralforecast#1337]
### What happened + What you expected to happen

When using Nixtla’s RNN-Direct or LSTM-Direct classes to forecast four steps at once, the resulting mean squared error is consistently identical across all horizons, which is far higher than expected and significantly worse than other models in the same framework. By contrast, non‐RNN direct‐forecasting models such as PatchTST(h=4) produce reasonable errors, and the recurrent versions (RNNRecurrent, LSTMRecurrent) also behave as expected. This discrepancy indicates that there is likely a bug in the multi‐step prediction logic for RNN‐based direct forecasting—when RNNs are tasked with predicting all horizons simultaneously, they appear to output implausibly high error values instead of matching the baseline performance of other models.

### Versions / Dependencies

neuralforecast: 3.0.1

### Reproduction script

import pandas as pd
from neuralforecast import NeuralForecast
from neuralforecast.models import LSTM

models = [LSTM(h=forecasting_horizon,  # forecasting_horizon > 1
                      input_size=look_back_size, 
                      max_steps=max_steps, 
                      learning_rate=learning_rate, 
                      loss=MSE(), random_seed=seed, 
                      scaler_type=scaler_type, 
                      batch_size=batch_size, 
                      step_size=step_size,
                      recurrent = False,                   
                     )]
nf = NeuralForecast(models=models, freq=1, local_scaler_type=scaler_type)
nf.fit(df_train, verbose=True, val_size=0)
preds = nf.predict(df_test)

### Issue Severity

High: It blocks me from completing my task.

**Comment by marcopeix** [comment]
Hello,

I tested the LSTM model with `recurrent=False` on the AirPassengers dataset, varying the horizon from 1 to 12, and I do not get an identical MSE for all horizons. 

```python
Y_train_df = AirPassengersPanel_train = AirPassengersPanel[AirPassengersPanel['ds'] < AirPassengersPanel['ds'].values[-12]].reset_index(drop=True)
Y_test_df = AirPassengersPanel_test = AirPassengersPanel[AirPassengersPanel['ds'] >= AirPassengersPanel['ds'].values[-12]].reset_index(drop=True)

horizons = range(1, 13)

for horizon in horizons:

    lstm = LSTM(
        h=horizon,
        input_size=24,
        recurrent=False,
        max_steps=200
    )
    nf = NeuralForecast(models=[lstm], freq="ME")
    nf.fit(df=Y_train_df)
    preds = nf.predict()
    test_df = pd.merge(Y_test_df, preds, 'left', ['unique_id', 'ds'])
    evaluation = evaluate(
        test_df,
        metrics=[mse],
        models=["LSTM"],
    )
    print(evaluation)
```

Otherwise, I think I'm missing the "bug" portion of this issue. 

1. The fact that the error is higher than expected (what is the expected error) might indicate that LSTM or RNN is not suitable for your dataset.
2. Models within the same framework will perform differently, that's normal. PatchTST is a very solid model that tends to perform well. 
3. Setting `recurrent=True` or `False` will change the performance of the model, since the predictions are generated differently. 

Let me know if I'm missing something, or if you have reproducible code with some data.

**Comment by alireza-jafari** [comment]
Thank you for your quick response!
By “multi-step forecasting,” I mean predicting multiple future time-steps at once, where the step_size is 1. It means we are forecasting 4 time series with horizon = 4,  the model outputs four series of forecasts: one step ahead (h₁), two steps ahead (h₂), three steps ahead (h₃), and four steps ahead (h₄).

I evaluated the following models—data scaled with scikit-learn’s StandardScaler:
models = [
    NBEATS(input_size=look_back_size, h=horizon, max_steps=200, scaler_type=None),
    PatchTST(input_size=look_back_size, h=horizon, max_steps=200, scaler_type=None),
    LSTM(input_size=look_back_size, h=horizon, max_steps=500, recurrent=False, scaler_type=None),
    LSTM(input_size=look_back_size, h=horizon, max_steps=500, recurrent=True, scaler_type=None)
]

**The behavior of the RNN-based models is not normal.** I tested it on AirPassengersDF, and please look at the results.

MSE on Prediction Length 1-4
| Horizon | NBEATS | PatchTST | LSTM‑direct | LSTM‑recurr |
|---------|-------:|---------:|------------:|---------------:|
| **h1**  | 0.025  | 0.020 | 0.459 | 0.720 |
| **h2**  | 0.031  | 0.024 | 0.396 | 0.727 |
| **h3**  | 0.034  | 0.028 | 0.360 | 0.739 |
| **h4**  | 0.039  | 0.024 | 0.325 | 0.758 |


We expected to see a pattern of rising error when the horizon lengthens.

![Image](https://github.com/user-attachments/assets/131402aa-4777-4385-99db-233b44353a8b)

**Comment by elephaint** [comment]
- You should always set a scaler when training an RNN. Try with `robust`.
- Train more steps, for recurrent easily 2000.

**Comment by alireza-jafari** [comment]
Thanks for the pointers!
I actually did scale the data before fitting, using sklearn.preprocessing.StandardScaler. To avoid double-scaling I kept scaler_type=None inside the model.
I pushed max_steps to 2000 as suggested. Convergence looks smoother, but the horizon-specific errors still don't make sense. 
I implemented a very simple LSTM in TensorFlow or PyTorch (about 20 lines of code) and compared its results with neuralforecast’s LSTM. The behavior of the neuralforecast’s RNN-based models still seems off. However, other models work perfectly well.

**Comment by marcopeix** [comment]
Hi @alireza-jafari . Can you share a script reproducing your results?

**Comment by alireza-jafari** [comment]
<html><head></head><body><p>Thanks for looking into this. I uploaded a minimal reproducible notebook here:</p><p><a href="https://drive.google.com/file/d/1DiEO2ISnugv_QNXnaNte01K73RAZBSDP/view?usp=sharing">https://drive.google.com/file/d/1DiEO2ISnugv_QNXnaNte01K73RAZBSDP/view?usp=sharing</a></p>

The notebook compares three models under the same rolling-origin evaluation protocol:

1. `PatchTST`
2. Nixtla `LSTM(recurrent=True)`
3. A simple TensorFlow/Keras recursive LSTM baseline

The Keras baseline is not intended to be architecture-identical to Nixtla’s LSTM. I included it as a sanity check for the recursive evaluation setup. It is trained as a one-step-ahead LSTM using the previous 12 observations, and during evaluation it predicts 4 steps recursively by feeding each prediction back into the input window.



| model                | horizon |      MSE |      MAE |     NNSE |        R2 |
| -------------------- | ------: | -------: | -------: | -------: | --------: |
| LSTM_Recurrent       |       1 | 0.883327 | 0.706064 | 0.507257 |  0.028612 |
| LSTM_Recurrent       |       2 | 0.743678 | 0.678893 | 0.541729 |  0.154058 |
| LSTM_Recurrent       |       3 | 0.770093 | 0.687696 | 0.519139 |  0.073735 |
| LSTM_Recurrent       |       4 | 0.892746 | 0.771137 | 0.475361 | -0.103666 |
| PatchTST             |       1 | 0.040078 | 0.161723 | 0.957787 |  0.955927 |
| PatchTST             |       2 | 0.057119 | 0.185414 | 0.938990 |  0.935026 |
| PatchTST             |       3 | 0.072045 | 0.208294 | 0.920255 |  0.913344 |
| PatchTST             |       4 | 0.075970 | 0.230985 | 0.914145 |  0.906081 |
| Keras_LSTM_Recurrent |       1 | 0.629778 | 0.627552 | 0.590820 |  0.307438 |
| Keras_LSTM_Recurrent |       2 | 0.816581 | 0.725237 | 0.518438 |  0.071130 |
| Keras_LSTM_Recurrent |       3 | 0.884303 | 0.765233 | 0.484582 | -0.063636 |
| Keras_LSTM_Recurrent |       4 | 0.941662 | 0.799786 | 0.462077 | -0.164139 |


<p>PatchTST behaves as expected: the error increases gradually as the forecast horizon increases.</p>

</code></pre><p>My expectation for a recurrent/recursive multi-step forecast is that horizon 1 should usually be the easiest prediction, and later horizons should generally become harder because later predictions depend on previous predictions. I understand that horizon-wise errors do not have to be perfectly monotonic in every dataset, especially with a small test set, but this result looks suspicious.</p>

**Comment by marcopeix** [comment]
@alireza-jafari 

Alright, I think I understand the discrepancy. In your Keras code, the state is rebuilt at each prediction step, so it's input_size + last prediction. 

In neuralforecast, we don't do that. We use the previously saved hidden state and only add the last prediction. 

So, you're using what we call a **window-based rollout**, and neuralforecast uses a **state-based rollout**. Both approaches are correct, but will give different results. With a state-based rollout, we save on computation, so it makes the model faster.

That explains the differences between your Keras baseline and our model. Ultimately, the strategies we suggested to increase training steps and use a scaler when `recurrent=True` remain.

I hope this clarifies everything.

**Comment by alireza-jafari** [comment]
Thank you for the explanation. That clarifies the difference in performances. 

My remaining concern is about the behavior of LSTM(recurrent=True) itself in this simple rolling-origin experiment, the Nixtla recurrent LSTM still shows unexpected horizon-wise behavior. I expected the recurrent LSTM to at least perform best at horizon 1 or degrade more smoothly, since horizon 1 does not depend on previous generated predictions.

Is the state-based rollout strategy for recurrent=True documented somewhere? I did not find anything. It may be useful to mention explicitly, because many users may assume recursive forecasting means rebuilding the input window at each step.

Thanks again for the clarification.
